Fix sqswatcher.py's removeHost

chambm opened the pull request:
removeHost() currently requires 3 arguments. I don't know if slots should come from the message attributes here or whether it can just be 0. But without the parameter, I get an error when the cluster autoshrinks.

dougalb commented:
What scheduler are you running with sqswatcher? The slots parameter is only needed for addHost and none of the removeHost functions in the plugins expect more than two parameters.

chambm commented:
SLURM. This is the stack trace I got from sqswatcher:

> [ec2-user@ip-10-0-1-202 ~]$ sudo /usr/local/bin/supervisorctl tail sqswatcher
2016-06-07 16:13:34,995 INFO [sqswatcher:main] sqswatcher startup
2016-06-07 16:13:35,112 INFO [sqswatcher:pollQueue] eventType=autoscaling:EC2_INSTANCE_TERMINATE
2016-06-07 16:13:35,112 INFO [sqswatcher:pollQueue] instanceId=i-01f4419d
Traceback (most recent call last):
  File "/usr/lib64/python2.7/logging/__init__.py", line 859, in emit    msg = self.format(record)
  File "/usr/lib64/python2.7/logging/__init__.py", line 732, in format    return fmt.format(record)
  File "/usr/lib64/python2.7/logging/__init__.py", line 471, in format    record.message = record.getMessage()
  File "/usr/lib64/python2.7/logging/__init__.py", line 335, in getMessage    msg = msg % self.args
TypeError: not all arguments converted during string formatting
Logged from file sqswatcher.py, line 192
Traceback (most recent call last):
  File "/usr/local/bin/sqswatcher", line 11, in <module>
    sys.exit(main())
  File "/usr/local/lib/python2.7/site-packages/sqswatcher/sqswatcher.py", line 211, in main
    pollQueue(scheduler, q, t)
  File "/usr/local/lib/python2.7/site-packages/sqswatcher/sqswatcher.py", line 185, in pollQueue
    s.removeHost(hostname,cluster_user)
TypeError: removeHost() takes exactly 3 arguments (2 given)

dougalb commented:
Can you post /usr/local/lib/python2.7/site-packages/sqswatcher/plugins/slurm.py

chambm commented:
```
# Copyright 2013-2015 Amazon.com, Inc. or its affiliates. All Rights Reserved.
#
# Licensed under the Amazon Software License (the "License"). You may not use this file except in compliance with the
# License. A copy of the License is located at
#
# http://aws.amazon.com/asl/
#
# or in the "LICENSE.txt" file accompanying this file. This file is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES
# OR CONDITIONS OF ANY KIND, express or implied. See the License for the specific language governing permissions and
# limitations under the License.

import subprocess as sub
from tempfile import mkstemp
from shutil import move
import os
import paramiko
import socket
import time
import logging

log = logging.getLogger(__name__)


def __runCommand(command):
    _command = command
    log.debug(repr(command))
    try:
        sub.check_call(_command, env=dict(os.environ))
    except sub.CalledProcessError:
        log.error("Failed to run %s\n" % _command)


def __restartSlurm(hostname, cluster_user):
    # Connect and restart Slurm on compute node
    ssh = paramiko.SSHClient()
    ssh.set_missing_host_key_policy(paramiko.AutoAddPolicy())
    hosts_key_file = os.path.expanduser("~" + cluster_user) + '/.ssh/known_hosts'
    user_key_file = os.path.expanduser("~" + cluster_user) + '/.ssh/id_rsa'
    iter=0
    connected=False
    while iter < 3 and not connected:
        try:
            log.info('Connecting to host: %s iter: %d' % (hostname, iter))
            ssh.connect(hostname, username=cluster_user, key_filename=user_key_file)
            connected = True
        except socket.error, e:
            log.error('Socket error: %s' % e)
            time.sleep(10 + iter)
            iter = iter + 1
            if iter == 3:
                log.critical("Unable to connect to host")
                return
    try:
        ssh.load_host_keys(hosts_key_file)
    except IOError:
        ssh._host_keys_filename = None
        pass
    ssh.save_host_keys(hosts_key_file)
    command = "sudo sh -c \'/etc/init.d/slurm restart &> /tmp/slurmdstart.log\'"
    stdin, stdout, stderr = ssh.exec_command(command)
    ssh.close()


def __readNodeList():
    _config = "/opt/slurm/etc/slurm.conf"
    nodes = {}
    with open(_config) as slurm_config:
        for line in slurm_config:
            if line.startswith('#PARTITION'):
                partition = line.split(':')[1].rstrip()
                dummy_node = slurm_config.next()
                node_name = slurm_config.next()
                items = node_name.split(' ')
                node_line = items[0].split('=')
                if len(node_line[1]) > 0:
                    nodes[partition] = node_line[1].split(',')
                else:
                    nodes[partition] = []
    return nodes


def __writeNodeList(node_list):
    _config = "/opt/slurm/etc/slurm.conf"
    fh, abs_path = mkstemp()
    with open(abs_path,'w') as new_file:
        with open(_config) as slurm_config:
            for line in slurm_config:
                if line.startswith('#PARTITION'):
                    partition = line.split(':')[1].rstrip()
                    new_file.write(line)
                    dummy_node = slurm_config.next()
                    new_file.write(dummy_node)
                    node_names = slurm_config.next()
                    partitions = slurm_config.next()
                    items = node_names.split(' ')
                    node_line = items[0].split('=')
                    if len(node_list[partition]) > 0:
                        new_file.write('NodeName=' + ','.join(node_list[partition]) + " " + ' '.join(items[1:]))
                    else:
                        new_file.write("#NodeName= Procs=1 State=UNKNOWN\n")
                    items = partitions.split(' ')
                    node_line = items[1].split('=')
                    new_file.write(items[0] + " " + node_line[0] + '=dummy-' + partition + ',' + ','.join(node_list[partition]) + " " + ' '.join(items[2:]))
                else:
                    new_file.write(line)
    os.close(fh)
    #Remove original file
    os.remove(_config)
    #Move new file
    move(abs_path, _config)
    #Update permissions on new file
    os.chmod(_config, 0744)


def addHost(hostname, cluster_user, slots):
    log.info('Adding %s' % hostname)

    # Get the current node list
    node_list = __readNodeList()

    # Add new node
    node_list['compute'].append(hostname)
    __writeNodeList(node_list)

    # Restart slurmctl locally
    command = ['/etc/init.d/slurm', 'restart']
    __runCommand(command)

    # Restart slurmctl on host
    __restartSlurm(hostname, cluster_user)

    # Reconfifure Slurm, prompts all compute nodes to reread slurm.conf
    command = ['/opt/slurm/bin/scontrol', 'reconfigure']
    __runCommand(command)


def removeHost(hostname, cluster_user, slots):
    log.info('Removing %s', hostname)

    # Get the current node list
    node_list = __readNodeList()

    # Remove node
    node_list['compute'].remove(hostname)
    __writeNodeList(node_list)

    # Restart slurmctl
    command = ['/etc/init.d/slurm', 'restart']
    __runCommand(command)

    # Reconfifure Slurm, prompts all compute nodes to reread slurm.conf
    command = ['/opt/slurm/bin/scontrol', 'reconfigure']
    __runCommand(command)
```

chambm commented:
Yes, I see slurm.py was fixed on Apr 19. But I just started this cluster yesterday. Shouldn't it have gotten the latest cfncluster-node files?

dougalb commented:
What version of cfncluster-node is installed on your MasterServer? What version of CfnCluster are you using? This was fixed here https://github.com/awslabs/cfncluster-node/commit/24d5117ddcaacf3f255779c29db0d60e279962fb and included in the latest version of CfnCluster which is 1.2.1 and released on 2016-03-24.

chambm commented:
`grep "cfncluster" /var/log/*log | grep version`
> /var/log/cfn-init-cmd.log:2016-06-06 20:01:48,402 P2904 [INFO]      - install version 1.1.2 of package cfncluster-node
> /var/log/cfn-init.log:    - install version 1.1.2 of package cfncluster-node
> /var/log/cloud-init-output.log:+ export cfncluster_version=cfncluster-1.2.1
> /var/log/cloud-init-output.log:+ cfncluster_version=cfncluster-1.2.1
> /var/log/cloud-init-output.log:+ export cookbook_version=cfncluster-cookbook-1.2.1
> /var/log/cloud-init-output.log:+ cookbook_version=cfncluster-cookbook-1.2.1
> /var/log/cloud-init-output.log:+ installed_version=cfncluster-1.2.1

dougalb commented:
Looking at your stacks, you are running your own custom templates and stack. This means that you do not get updates automatically or notification of updates via the CLI. To get the latest version of cfncluster-node, please update your template or cookbooks.

chambm commented:
But doesn't this mean I have 1.2.1 (in the template)?

> "CfnClusterVersions" : {
>       "default" : {
>         "cfncluster" : "cfncluster-1.2.1",
>         "cookbook" : "cfncluster-cookbook-1.2.1",
>         "chef" : "12.8.1",
>         "ridley" : "4.5.0",
>         "berkshelf" : "4.3.0",
>         "ami" : "201603231645"
>       }
>     },

dougalb commented:
cfncluster-1.2.1 is using the following versions - https://github.com/awslabs/cfncluster/blob/master/cloudformation/cfncluster.cfn.json#L1087

chambm commented:
Ah. Just one character off! Ok, thanks. :) I'll have to come up with a script to update the template.
